README
About OSPD-DEBSECAN
-------------------

This is a OSP server implementation to allow GVM to remotely control
a debsecan scanner, see http://www.enyo.de/fw/software/debsecan/
OSPD-DEBSECAN identifies vulnerable packages on the Debian system it is
executed on.

Once running, you need to configure the Scanner for Greenbone Vulnerability
Manager, for example via the web interface Greenbone Security Assistant.
Then you can create scan tasks to use this scanner.

OSPD-DEBSECAN is licensed under GNU General Public License Version 2 or
any later version.  Please see file COPYING for details.

All parts of OSP-DEBSECAN are Copyright (C) by Greenbone Networks GmbH
(see http://www.greenbone.net).


How to start OSPD-DEBSECAN
--------------------------

There are no special usage aspects for this module
beyond the generic usage guide.

The only exception is that starting a scan requires some more
paramaters:

$ gvm-cli socket --sockpath <prefix>/var/run/ospd-debsecan.sock \
  --xml "<start_scan target='127.0.0.1' ports=''><scanner_params/></start_scan>"


Generic usage guide for OSPD-based scanners
-------------------------------------------

Replace "scanner" by the name of the actual OSPD scanner.

All OSPD servers share a set of command-line options such as
--help, --bind-address, --port, --key-file, --timeout etc.

For example to see the command line options you can run

$ ospd-scanner --help


To run a test instance of ospd-scanner on unix socket:

$ ospd-scanner -u <prefix>/var/run/ospd-scanner.sock &

To run a test instance of ospd-scanner on local port 1234:

$ ospd-scanner -b 127.0.0.1 -p 1234 &

Add "--log-level=DEBUG" to enable maximum debugging output.

Parameter for --log-level can be one of debug, info, warnings, error, or
critical (in order of priority).


Controlling a OSP daemon
------------------------

You can use command line tools of "gvm-tools" module to send to and
receive information from the scanners.

Get a description of the interface:

$ gvm-cli socket --sockpath <prefix>/var/run/ospd-scanner.sock --xml "<help/>"

or

$ gvm-cli --use-certs -p 1234 -i --xml="<help/>"


Starting a scan (scanner parameters can be added according to the description
printed upon the command "<help/>":

$ gvm-cli socket --sockpath <prefix>/var/run/ospd-scanner.sock --xml="<start_scan target='www.example.com'></start_scan>"


Show the list of scans, their status and results:

$ gvm-cli socket --sockpath <prefix>/var/run/ospd-scanner.sock --xml="<get_scans/>"


CHANGES
SUMMARY OF RELEASE CHANGES FOR OSPD-DEBSECAN
============================================

For detailed code changes, please visit
 https://github.com/greenbone/ospd-debsecan/commits/master
or get the entire source code repository and view log history:
$ git clone https://github.com/greenbone/ospd-debsecan.git
$ cd ospd-debsecan && git log


ospd-debsecan 1.2b1 (2016-11-16)

This is the first beta release of the 1.2 version of the OSP (OpenVAS Scanner
Protocol) server for the debsecan tool to allow the retrieval of the debsecan
scan results into the Open Vulnerability Assessment System (OpenVAS).

Many thanks to everyone who has contributed to this release:
Benoît Allard, Hani Benhabiles and Jan-Oliver Wagner.

Main changes since 1.0.0:
* All alarms now have an explicit QoD value of 97%.
* Support for multi-host scanning has been added.
* A test-suite has been added.
* The documentation has been updated.


osp-debsecan 1.0.0 (2015-06-01)

This is the 1.0.0 release of the OSP (OpenVAS Scanner Protocol) server
for the debsecan tool to allow the retrieval of the debsecan scan results
into the Open Vulnerability Assessment System (OpenVAS).

This release supports calling a local debsecan installation.
About debsecan, see http://www.enyo.de/fw/software/debsecan/
 
Central features of osp-debsecan are:

* Call local debsecan tool and extract all CVE vulnerabilities.

osp-debsecan is Free Software (Open Source), licensed under GNU General Public License
Version 2 or any later version.  

Many thanks to everyone who has contributed to this release:
Jan-Oliver Wagner.


tests/testProcessOutput.py
# -*- coding: utf-8 -*-
# Description:
# Test-Suite for the debsecan processing
#
# Authors:
# Benoît Allard <benoit.allard@greenbone.net>
#
# Copyright:
# Copyright (C) 2015 Greenbone Networks GmbH
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.

import unittest

from ospd_debsecan.wrapper import process_output

class testProcessOutput(unittest.TestCase):

    def testCVENoComment(self):
        res = list(process_output(["CVE-2014-5044 gcc-4.6"]))
        self.assertEqual(len(res), 1)
        self.assertEqual(res[0], dict(name='CVE-2014-5044', value='CVE-2014-5044 gcc-4.6'))

    def testCVEWithComment(self):
        res = list(process_output(["CVE-2002-2439 gcc-4.6 (low urgency)"]))
        self.assertEqual(len(res), 1)
        self.assertEqual(res[0], dict(name='CVE-2002-2439', value='CVE-2002-2439 gcc-4.6 (low urgency)'))

    def testNoCVE(self):
        res = list(process_output(["TEMP-0000000-EA424A libbluray1"]))
        self.assertEqual(len(res), 0)
       


ospd_debsecan/wrapper.py
# -*- coding: utf-8 -*-
# Description:
# Setup for the OSP debsecan Server
#
# Authors:
# Jan-Oliver Wagner <Jan-Oliver.Wagner@greenbone.net>
#
# Copyright:
# Copyright (C) 2015 Greenbone Networks GmbH
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.

from ospd.ospd_ssh import OSPDaemonSimpleSSH
from ospd.misc import main as daemon_main
from ospd_debsecan import __version__

OSPD_DESC = """
This scanner runs the tool 'debsecan' (Debian Security Analyzer) on the target hosts via a ssh connection.

This tool is available for Debian GNU/Linux systems and can create a list of vulnerabilities the present system is subject to.
The tool goes beyond a simple match about which package updates are missing.
It rather does an online query to learn about software packages where security problems are identified, but not yet been finally processed to become a security update.

A QoD of 97% (QoD type 'package' for authenticated package-based checks) is applied to any results.

Further details about the tool are available on the debsecan homepage:

http://www.enyo.de/fw/software/debsecan/

For executing debsecan a low privileged user account is sufficient on the respective target systems.
Also the tool 'debsecan' needs to be installed on the target systems.
"""

def process_output(output):
    """ Generates values for the add_scan_alarm method """
    for line in output:
        if not line.startswith('CVE-'):
            continue
        yield dict(name=line.split()[0], value=line)


class OSPDdebsecan(OSPDaemonSimpleSSH):

    """ Class for ospd-debsecan daemon. """

    def __init__(self, certfile, keyfile, cafile):
        """ Initializes the ospd-debsecan daemon's internal data. """
        super(OSPDdebsecan, self).__init__(certfile=certfile, keyfile=keyfile,
                                            cafile=cafile)
        self.server_version = __version__
        self.scanner_info['name'] = 'debsecan'
        self.scanner_info['version'] = 'depends on the local installation at the target host'
        self.scanner_info['description'] = OSPD_DESC

    def check(self):
        """
        Checks that debsecan command line tool is found and is executable.
        Since this osp scanner uses local installations on many systems
        where on some systems it might be available and on others not be present,
        we simply return True.
        """

        return True

    def exec_scan(self, scan_id, target):
        """ Starts the debsecan scanner for scan_id scan. """

        result = self.run_command(scan_id=scan_id, host=target, cmd="debsecan")

        if result is None:
            self.add_scan_error(scan_id, host=target,
              value="A problem occurred trying to execute 'debsecan'.")
            return 2

        for alarm in process_output(result):
            self.add_scan_alarm(scan_id, host=target, qod='97', **alarm)
        return 1

def main():
    """ OSP debsecan main function. """
    daemon_main('OSPD - debsecan wrapper', OSPDdebsecan)
